### PR TITLE
Flickr add upload method

### PR DESCRIPTION
**Pull request by kylewm**
- Flickr uploads uses a separate API from other requests
- use requests.post for upload (urllib2 doesn't help with multi-part)
- If no permission scopes are requested, use the authenticate
  API instead of authorize (returns without prompting if previous authed)

### CODE REVIEW

**Comment by snarfed** [comment]
cool!

should raise_for_failure maybe go into webutil.util.interpret_http_exception?

I'm also curious to hear the rationale for why upload is here instead of granary. i get that it's a different api, but still...?

**Comment by snarfed** [comment]
also, XML, whee!

**Comment by kylewm** [comment]
It does seem more consistent for the Flickr error code interpretation in `raise_for_failure` to be in `interpret_http_exception` instead; I'm not sure how to actually move it though. Would I raise a 400 error with the message and error code in the body of the error, and then in `interpret`, upgrade it to 401 based on the error body?

My thinking on `flickr_auth.upload` was that it's basically a special case of `flickr_auth.call_api_method`, and I like having the OAuth1 machinations handled all in this one module. Then the Granary implementation is all about logic and not about manipulating HTTP headers or whatever. That said, I don't feel strongly -- happy to move to Granary if you'd prefer it.

File: oauth_dropins/flickr_auth.py [hunk 1]
```diff
@@ -72,13 +86,76 @@ def call_api_method(method, params, token_key, token_secret):
...
-    else:
-      code = 400
-    raise urllib2.HTTPError(url, code, 'message=%s, flickr code=%d' % (
-      body.get('message'), error_code), {}, None)
+    raise_for_failure(url, body.get('code'), body.get('message'))
 
   return body
+
+
+def upload(params, photo_file, token_key, token_secret, timeout=None):
```
Comment (snarfed) [inline oauth_dropins/flickr_auth.py]: re timeout, yeah, maybe just remove it here and hard-code `appengine_config.HTTP_TIMEOUT`? that's what we do elsewhere.

**Comment by snarfed** [comment]
oauth 1! right. makes sense, sgtm.

and yeah, you're right, `interpret_http_exception` isn't such a great fit after all, since there's no exception here. it's fine as is.

**Pull request merged**